clients/src/main/java/org/apache/kafka/common/utils/CollectionUtils.java

```diff
@@ -67,13 +66,29 @@ private CollectionUtils() {}
         Map<String, List<Integer>> partitionsByTopic = new HashMap<>();
         for (TopicPartition tp: partitions) {
             String topic = tp.topic();
-            List<Integer> topicData = partitionsByTopic.get(topic);
-            if (topicData == null) {
-                topicData = new ArrayList<>();
-                partitionsByTopic.put(topic, topicData);
-            }
+            List<Integer> topicData = partitionsByTopic.computeIfAbsent(topic, k -> new ArrayList<>());
             topicData.add(tp.partition());
         }
         return  partitionsByTopic;
     }
+
+    /**
+     * Abstract duplicate code from StickAssignor and ConsumerProtocol
+     *
+     * @param struct                 data source
+     * @param topicPartitionsKeyName Topic Partitions key names
+     * @return list of TopicPartition from struct
+     */
+    public static List<TopicPartition> getTopicPartitionFromStruct(Struct struct, String topicPartitionsKeyName) {
+        List<TopicPartition> partitions = new ArrayList<>();
```

Comment:
You could set the size explicitly using the array from the struct.

Reply:
Setting size explicitly for List is not a better way to initialize List, because we don't have much idea about this list will be use for how, its actual size is not clear.

Reply:
Yes, my bad. The top-level array is not sufficient to determine the size.